### PR TITLE
Port command line options from SDL executable [SDL PR 1/2]

### LINKED ISSUE #462: Port SDL executable command line functionality to Qt executable
### Issue

_No response_

### PR DESCRIPTION
**Pull request by OpenSauce04**
Closes #462

This is the first of two PRs which will port functionality from the old SDL executable and merge it with the Qt executable.
This one focuses on porting the command line options which are available via the SDL executable.

This list keeps track of the options as they are implemented:
- [x] Dump video `-d`
- [x] GDB port functionality `-g`
  - [x] Remove old `-g` option, allow for a file to be provided at the end of the command
- [x] Show help `-h`
- [x] Install CIA from the command line `-i`
- [x] Play TAS movie `-p`
- [x] Record TAS movie `-r`
- [x] Show version `-v`

### CODE REVIEW

**Comment by OpenSauce04** [comment]
Any platform-specific compilation errors will be addressed once the pull request is feature-complete

File: src/lime_qt/main.cpp [hunk 1]
```diff
@@ -16,6 +17,8 @@
 #include <QtWidgets>
 #include <fmt/format.h>
 #include <fmt/ostream.h>
+#undef _UNICODE
```
Comment (DavidRGriswold) [inline src/lime_qt/main.cpp]: Why is this necessary? Not doubting, but my limited understanding of #undef makes me think it's something we should be sure of

**Review by DavidRGriswold — COMMENTED** [review]
Okay, just making a checklist for myself for testing here to make sure everything is tested...

- [x] Test -h, -v, on mac (easy)
   - [x] -f not working yet, see error below
- [x] Test -d on mac (do I even have ffmpeg?)
- [x] Test -i on mac
- [x] Test -g on mac (first ask how to connect to a gdb stub)
- [x] Test the TAS commands on mac
- [ ] Repeat all of the above on windows

**Review by DavidRGriswold — COMMENTED** [review]
Welp, on my m2 Macbook air the `-f` tag seems to be broken.

I can launch a game no problem with `lime3ds path_to_rom` and it uses whatever the prior view setting was.

If I am playing a game windowed, quit, and I try to run `lime3ds -f path_to_rom` , I get a segmentation fault. 

If I am playing a game fullscreen, quit while it's still fullscreen, and try to run `lime3ds -f path_to_rom` it launches the main UI but does not start the game

**Comment by DavidRGriswold** [comment]
The error message when you don't have ffmpeg installed takes us to the internet archive of the old citra wiki still, here https://web.archive.org/web/20240301121456/https://citra-emu.org/wiki/installing-ffmpeg-for-the-video-dumper/ . Not sure if we want to move that help file to a lime3ds spot, remove the help link at all, or leave it.

Also, those instructions don't seem to work anyway. I installed ffmpeg@6 from homebrew as instructed, added ffmpeg to my path, and am still not getting it to work on mac.

That said, this is an error that has more to do with the Video Dumper Tool in its entirety and should probably be reported as a separate bug, since the same issue happens if I do not use the command line argument. So I'm going to assume the PR itself is working as correctly and maybe make an issue for this bug later.

**Comment by DavidRGriswold** [comment]
Hmm, I don't think `-g` is compatible with modern macs. GDB doesn't run on arm macs. I guess it's possible the code will work with clang or other debuggers, but I am not sure how to test that.

**Comment by DavidRGriswold** [comment]
None of the interactive ones (h, v) are working on windows under cmd or powershell. More testing to come. 

I went back to 2116 and confirmed that -h, at least, works on the lime3ds-cli.exe

File: src/lime_qt/main.cpp (hunk 1 above)
Comment (OpenSauce04) [reply src/lime_qt/main.cpp]: Without this Windows compilation completely flops due to complications with `wchar_t` nonsense. This same undef was also used in the SDL frontend.

**Comment by OpenSauce04** [comment]
> Welp, on my m2 Macbook air the -f tag seems to be broken.

Fixed by e90c3f2

**Comment by ghost** [comment]
I also confirm the windows build basically is completely broken

The TAS args and -g seem to jsut simply open the executable, while the rest do nothing

**Comment by OpenSauce04** [comment]
I've addressed the Windows command line issues by displaying the command output as a Qt popup on that platform.

I did a lot of research to try and find ways that don't involve doing this, but it seems that Windows is designed in a way that is fundamentally incompatible with hybrid CLI-GUI applications. The method I chose seemed to get the job done without fighting against the Windows API in an exceedingly hacky fashion.

Despite this, I do not want to go back to the old hard executable split between SDL and Qt, as I strongly believe it is the wrong approach. This somewhat unconventional delivery of information is a compromise I'm fine with.

**Comment by OpenSauce04** [comment]
There is some extra refactoring I want to make to this PR before merging to improve responsiveness by moving the argument checking to earlier in the code, but it's too late at night for me to get into that now.

Outside of that the code should be in a reviewable state. Apologies for the force-push spam, I kept finding minor things I missed during the message box change. It's been a long day.

**Comment by rtiangha** [comment]
I know a lot of work has been put in already to get the functionality to where it is, but if people were still interested in figuring out how to pipe the stdout/stderr messages to a Windows terminal window, the code found on [this page](https://www.tillett.info/2013/05/13/how-to-create-a-windows-program-that-works-as-both-as-a-gui-and-console-application/) (which is MIT licensed so should be GPL2 compatible) works well enough. Just toss it into a `.h` file, include it in `main.cpp` and link it in via cmake, then call `attachOutputToConsole()` at the start of `main()` like in the example main function, and call the hacky `sendEnterKey()` function before returning from main wherever you call return (it explains why you need to do it that way in the attached blog post) and it should just go.

I've been playing around with it on my end and it seems to work well enough; the only thing it doesn't work with by default is a mintty based terminal, but even that can be worked around by looking at the `TERM` system variable before invoking and if it's set to `xterm`, then just ignore and keep going because msys2 and similar terminals handle the stdout messages just fine as is.

**Comment by DavidRGriswold** [comment]
I have tested `-i`, `-h`, `-f`, and `-v` on windows and all work as expected with the message boxes. I should hopefully have time to test the others today.

**Comment by DavidRGriswold** [comment]
`-p` and `-r` work fine on windows too. Struggling to test `-d` because archive.org is down so I don't know which version of ffmpeg to install - the version from winget only provides binaries and lime complains that it needs libavcodec.dll.

**Review by DavidRGriswold — APPROVED** [review]
Code reviewed and I have tested everything I can on Windows and Mac. FFMpeg issues mean i have not tested -d fully but it pops up the correct error box, so it seems to at least be doing the same thing as the manual choice in the UI to dump the video, which is enough to approve the PR in my opinion.

Code all looks fine, no issues that are obvious to me.

**Comment by OpenSauce04** [comment]
@rtiangha in the interest of merging this in a timely fashion, I will consider that to be a separate matter. Please feel free to open an issue if you wish to discuss it further.
